### ISSUE TITLE
Panic in pool: invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by FZambia** [go-redis/redis#262]
Hello. Experimenting with go-redis in [Centrifugo](https://github.com/centrifugal/centrifugo) - got panic after testing work when simulating Redis server hanging:

```
redis-cli DEBUG sleep 30
```

Panic:

```
[E]: 2016/03/02 01:21:32 read tcp 127.0.0.1:49650->127.0.0.1:6379: i/o timeout
redis: 2016/03/02 01:21:33 Unwatch failed: read tcp 127.0.0.1:49646->127.0.0.1:6379: i/o timeout
[C]: 2016/03/02 01:21:33 read tcp 127.0.0.1:49651->127.0.0.1:6379: i/o timeout
2016/03/02 01:21:33 http: panic serving [::1]:64495: runtime error: invalid memory address or nil pointer dereference
goroutine 6 [running]:
net/http.(*conn).serve.func1(0xc8200c3760, 0xb97888, 0xc82002e180)
	/Users/fz/projects/go15/src/net/http/server.go:1287 +0xb5
gopkg.in/redis%2ev3.(*connPool).Remove(0xc8201087d0, 0xc820372ae0, 0x0, 0x0, 0x0, 0x0)
	/private/var/www/different/go/gopath/src/gopkg.in/redis.v3/pool.go:295 +0x34
gopkg.in/redis%2ev3.(*baseClient).putConn(0xc82010faa0, 0xc820372ae0, 0x0, 0x0)
	/private/var/www/different/go/gopath/src/gopkg.in/redis.v3/redis.go:28 +0xb5
gopkg.in/redis%2ev3.(*baseClient).process(0xc82010faa0, 0x1b055c0, 0xc8202ea780)
	/private/var/www/different/go/gopath/src/gopkg.in/redis.v3/redis.go:71 +0x375
gopkg.in/redis%2ev3.(*baseClient).(gopkg.in/redis%2ev3.process)-fm(0x1b055c0, 0xc8202ea780)
	/private/var/www/different/go/gopath/src/gopkg.in/redis.v3/redis.go:189 +0x34
gopkg.in/redis%2ev3.(*commandable).Process(0xc82015c338, 0x1b055c0, 0xc8202ea780)
	/private/var/www/different/go/gopath/src/gopkg.in/redis.v3/commands.go:57 +0x34
gopkg.in/redis%2ev3.(*commandable).EvalSha(0xc82015c338, 0xc82015a480, 0x28, 0xc8201fd950, 0x1, 0x1, 0xc820373620, 0x6, 0x6, 0x0)
	/private/var/www/different/go/gopath/src/gopkg.in/redis.v3/commands.go:1582 +0x52e
gopkg.in/redis%2ev3.(*Script).EvalSha(0xc82010fae0, 0x1b05580, 0xc82015c330, 0xc8201fd950, 0x1, 0x1, 0xc820373620, 0x6, 0x6, 0x18c)
	/private/var/www/different/go/gopath/src/gopkg.in/redis.v3/script.go:43 +0x99
gopkg.in/redis%2ev3.(*Script).Run(0xc82010fae0, 0x1b05580, 0xc82015c330, 0xc8201fd950, 0x1, 0x1, 0xc820373620, 0x6, 0x6, 0x23f36)
	/private/var/www/different/go/gopath/src/gopkg.in/redis.v3/script.go:47 +0x98
github.com/centrifugal/centrifugo/libcentrifugo.(*RedisEngine).publish(0xc820108820, 0xc8204f0a00, 0x1f, 0xc820442c00, 0x1a8, 0x2e5, 0xc82020a540, 0x0, 0x0)
	/private/var/www/different/go/gopath/src/github.com/centrifugal/centrifugo/libcentrifugo/engineredis.go:489 +0x650
github.com/centrifugal/centrifugo/libcentrifugo.(*Application).pubClient(0xc820156140, 0xc8201fd400, 0xc, 0x101010101, 0xa, 0x1e, 0x1, 0xc8201fd420, 0xd, 0x10, ...)
	/private/var/www/different/go/gopath/src/github.com/centrifugal/centrifugo/libcentrifugo/application.go:456 +0x5a4
github.com/centrifugal/centrifugo/libcentrifugo.(*Application).publish(0xc820156140, 0xc8201fd400, 0xc, 0xc8201fd420, 0xd, 0x10, 0xc8201f4540, 0x24, 0xc8204ac660, 0xc820222501, ...)
	/private/var/www/different/go/gopath/src/github.com/centrifugal/centrifugo/libcentrifugo/application.go:408 +0x2ac
github.com/centrifugal/centrifugo/libcentrifugo.(*client).publishCmd(0xc8200f4c30, 0xc8204ac630, 0x30, 0x0, 0x0)
	/private/var/www/different/go/gopath/src/github.com/centrifugal/centrifugo/libcentrifugo/client.go:924 +0x3b9
github.com/centrifugal/centrifugo/libcentrifugo.(*client).handleCmd(0xc8200f4c30, 0xc8201fd3d0, 0x2, 0xc8201fd3a0, 0x7, 0xc8204ac600, 0x2f, 0x30, 0x449a00, 0x0, ...)
	/private/var/www/different/go/gopath/src/github.com/centrifugal/centrifugo/libcentrifugo/client.go:456 +0x846
github.com/centrifugal/centrifugo/libcentrifugo.(*client).handleCommands(0xc8200f4c30, 0xc8201182a0, 0x1, 0x4, 0x0, 0x0)
	/private/var/www/different/go/gopath/src/github.com/centrifugal/centrifugo/libcentrifugo/client.go:393 +0x206
github.com/centrifugal/centrifugo/libcentrifugo.(*client).message(0xc8200f4c30, 0xc8204e8c00, 0x5a, 0x600, 0x0, 0x0)
	/private/var/www/different/go/gopath/src/github.com/centrifugal/centrifugo/libcentrifugo/client.go:358 +0x74a
github.com/centrifugal/centrifugo/libcentrifugo.(*Application).RawWebsocketHandler(0xc820156140, 0xb97ac8, 0xc8200c3810, 0xc8200e81c0)
	/private/var/www/different/go/gopath/src/github.com/centrifugal/centrifugo/libcentrifugo/handlers.go:275 +0x5c0
github.com/centrifugal/centrifugo/libcentrifugo.(*Application).RawWebsocketHandler-fm(0xb97ac8, 0xc8200c3810, 0xc8200e81c0)
	/private/var/www/different/go/gopath/src/github.com/centrifugal/centrifugo/libcentrifugo/handlers.go:92 +0x3e
net/http.HandlerFunc.ServeHTTP(0xc820193380, 0xb97ac8, 0xc8200c3810, 0xc8200e81c0)
	/Users/fz/projects/go15/src/net/http/server.go:1422 +0x3a
github.com/centrifugal/centrifugo/libcentrifugo.(*Application).WrapShutdown.func1(0xb97ac8, 0xc8200c3810, 0xc8200e81c0)
	/private/var/www/different/go/gopath/src/github.com/centrifugal/centrifugo/libcentrifugo/handlers.go:470 +0xb8
net/http.HandlerFunc.ServeHTTP(0xc820190480, 0xb97ac8, 0xc8200c3810, 0xc8200e81c0)
	/Users/fz/projects/go15/src/net/http/server.go:1422 +0x3a
github.com/centrifugal/centrifugo/libcentrifugo.(*Application).Logged.func1(0xb97ac8, 0xc8200c3810, 0xc8200e81c0)
	/private/var/www/different/go/gopath/src/github.com/centrifugal/centrifugo/libcentrifugo/handlers.go:486 +0x13f
net/http.HandlerFunc.ServeHTTP(0xc8201904a0, 0xb97ac8, 0xc8200c3810, 0xc8200e81c0)
	/Users/fz/projects/go15/src/net/http/server.go:1422 +0x3a
net/http.(*ServeMux).ServeHTTP(0xc82015b8f0, 0xb97ac8, 0xc8200c3810, 0xc8200e81c0)
	/Users/fz/projects/go15/src/net/http/server.go:1699 +0x17d
net/http.serverHandler.ServeHTTP(0xc820110f00, 0xb97ac8, 0xc8200c3810, 0xc8200e81c0)
	/Users/fz/projects/go15/src/net/http/server.go:1862 +0x19e
net/http.(*conn).serve(0xc8200c3760)
	/Users/fz/projects/go15/src/net/http/server.go:1361 +0xbee
created by net/http.(*Server).Serve
	/Users/fz/projects/go15/src/net/http/server.go:1910 +0x3f6
```

This could be my mistake in the end, but anyway it should not result in panic.

**Comment by vmihailenco** [comment]
Hello. I am 99% sure this panic was fixed yesterday in https://github.com/go-redis/redis/pull/261. So please try to upgrade.

But overall it looks like there can be another issue, because there is unread data left in the connection. Is there any way I can try to reproduce this issue? Alternatively please look in the log for `connection has unread data: %q` message. It can reveal something interesting.

**Comment by FZambia** [comment]
@vmihailenco I was able to reproduce this only once since then - with the same looking panic. But I have no ideas how to reproduce this reliably - it just happens occasionally. I am using version without yesterday's fix. So hopefully this was fixed. I'll try to find a way to reproduce this so we can be sure it was fixed - no ideas at moment though. Or will try to use version with fix and maybe panic will be there too.

After Redis hanging goes away I usually get lots of log messages containing `connection has unread data...` - but when panic happened there were no such messages around. Is this ok to have such messages? Or such log messages should never appear in normal workflow?

**Comment by vmihailenco** [comment]
>Or will try to use version with fix and maybe panic will be there too.

I am sure I understand what caused that panic. So you should not see the *same* panic again :)

>Or such log messages should never appear in normal workflow?

I think they should not appear in normal workflow. At least I don't see them in production logs.

But now I understand what is happening. When you use `redis-cli DEBUG sleep 30` Redis Server does not answer immediately, but instead buffers commands and after 30 seconds sends replies for all buffered commands. But go-redis expects answer for only one command so it closes the connection and opens new one. That is okay, but probably go-redis should always close connection on timeout? I am not sure, because timeout is temporal error in Go (net.Error.Temporary()) and it did not caused any problems in practise.

**Comment by FZambia** [comment]
> So you should not see the same panic again :)

Ok thanks:) you can close this then to not confuse other users - I'll reopen it if I get the same panic with new version.

I also have some other questions:

1) I see that method that allowed turning connection into PubSub connection [is now deprecated](https://godoc.org/gopkg.in/redis.v3#Client.PubSub). I don't know the history behind this - but it seems to be useful shortcut when we don't need to subscribe immediately and just want to start subscribing later.
2) Seems that when I `publish` message into channel during Redis hanging and I have `MaxRetries` option set to something > 0 then this results in publishing message several times when Redis becomes available.

**Comment by vmihailenco** [comment]
>but it seems to be useful shortcut when we don't need to subscribe immediately and just want to start subscribing later.

Why not to just `Subscribe` later? :) We can allow `Subscribe` with zero arguments if there is a use case.

>results in publishing message several times when Redis becomes available.

If you use `DEBUG sleep` then Redis buffers published messages and then makes them available all at once. But that is true that go-redis retries publish on timeout and `MaxRetries > 0`. So seems to work as expected.

**Comment by FZambia** [comment]
> Why not to just Subscribe later?

Because my code looks like this:

```go
func (e *RedisEngine) runPubSub() {
	pubsub := e.client.PubSub()
	defer pubsub.Close()

	done := make(chan struct{})
	defer close(done)

	// Run subscriber goroutine
	go func() {
		for {
			select {
			case <-done:
				return
			case chan := <-e.sub:
				err := pubsub.Subscribe(chan)
			}
            ...
		}
	}()
    ...
}
```
So it's just a bit more simple to initialise `pubsub` connection outside goroutine in upper scope and then just use that `pubsub` connection in goroutine.

**Comment by dim** [comment]
So I have a similar issue with a cluster pipeline:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x20 pc=0x5288fd]

goroutine 783769 [running]:
gopkg.in/redis%2ev3.(*connPool).Remove(0xc21a94be50, 0xc21bd70e40, 0x0, 0x0, 0x0, 0x0)
        /go/src/gopkg.in/redis.v3/pool.go:295 +0x3d
gopkg.in/redis%2ev3.(*baseClient).putConn(0xc216213860, 0xc21bd70e40, 0x0, 0x0)
        /go/src/gopkg.in/redis.v3/redis.go:28 +0xaa
gopkg.in/redis%2ev3.(*ClusterPipeline).Exec(0xc212935560, 0xc208075500, 0x130, 0x150, 0x0, 0x0)
        /go/src/gopkg.in/redis.v3/cluster_pipeline.go:82 +0x784
```

**Comment by vmihailenco** [comment]
Please try to upgrade. It panicked on `p.storeLastErr(reason.Error())`, because reason was nil in one of the code paths...
